primo/utils/clustering_utils.py

```diff
@@ -97,39 +101,67 @@ def distance_matrix(wd: WellData, weights: dict) -> np.ndarray:
     )
 
 
-def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
+def _well_clusters(wd: WellData) -> dict:
+    """Returns well clusters"""
+    col_names = wd.col_names
+    set_clusters = set(wd[col_names.cluster])
+    return {
+        cluster: wd.data[wd[col_names.cluster] == cluster].index.to_list()
+        for cluster in set_clusters
```

Comment:
This can be more efficiently achieved using pandas groupby with a suitable aggregation. Please change.